docs/docs/using_custom_query_parser.md

```diff
@@ -0,0 +1,119 @@
+
+## Custom Query Parameter Parser
+
+Create a custom [QueryParameterParser](https://github.com/grpc-ecosystem/grpc-gateway/blob/main/runtime/query.go#L30) to handle the nested PageOptions message.
+```go
+package customparser
+
+import (
+	"net/url"
+	"strconv"
+
+	"github.com/grpc-ecosystem/grpc-gateway/v2/runtime"
+	"github.com/grpc-ecosystem/grpc-gateway/v2/utilities"
+	"google.golang.org/protobuf/proto"
+
+	your_service_v1 "path/to/your/service/v1"
+)
+
+// CustomQueryParameterParser parses query parameters into the appropriate gRPC message fields.
+type CustomQueryParameterParser struct{}
+
+// Parse parses query parameters and populates the appropriate fields in the gRPC request message.
+func (p *CustomQueryParameterParser) Parse(target proto.Message, values url.Values, filter *utilities.DoubleArray) error {
+	switch req := target.(type) {
+	// Different messages/requests can have different parsers, of course
+	case *your_service_v1.ListStuffRequest:
+		return populateListStuffParams(values, req)
+	}
+	
+	return (runtime.DefaultQueryParser{}).Parse(target, values, filter)
+}
+
+// populateListStuffParams populates the ListStuffRequest with query parameters.
+func populateListStuffParams(values url.Values, r *your_service_v1.ListStuffRequest) error {
+	pageOptions := &your_service_v1.PageOptions{}
+	
+	if limit := values.Get("limit"); limit != "" {
+		if parsedLimit, err := strconv.Atoi(limit); err == nil {
+			pageOptions.Limit = int32(parsedLimit)
+		}
+	}
+	if page := values.Get("page"); page != "" {
+		if parsedPage, err := strconv.Atoi(page); err == nil {
+			pageOptions.Page = int32(parsedPage)
+		}
+	}
+
+	r.Pagination = pageOptions
+	return nil
+}
+```
+## Integrate Your Custom Query Parameter Parser in gRPC-Gateway Setup
+
+All you need to do now is update the gRPC-Gateway setup, particularly wherever you are [defining the mux server](https://github.com/grpc-ecosystem/grpc-gateway/blob/main/runtime/mux.go#L293), to [use the custom query parameter parser](https://github.com/grpc-ecosystem/grpc-gateway/blob/main/runtime/mux.go#L110).
+```go
+package main
+
+import (
+	"context"
+	"github.com/grpc-ecosystem/grpc-gateway/v2/runtime"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/protobuf/encoding/protojson"
+	"net/http"
+
+	"your_module/path/customparser"
+)
+
+// create a new ServeMux with custom parser and other runtime options as needed
+func createGRPCGatewayMux() *runtime.ServeMux {
+	// whatever custom code you may need before you create the mux...
+	
+	return runtime.NewServeMux(
+		// Custom query parameter parser
+		runtime.WithQueryParameterParser(&customparser.CustomQueryParameterParser{}),
```

Comment:
I don't think this option exists, does it? You have to use `SetQueryParameterParser`: https://pkg.go.dev/github.com/grpc-ecosystem/grpc-gateway/v2/runtime#SetQueryParameterParser. Note that this will set the query parameter parser globally. It's honestly a bit gross, but it's the way it is because the parse function is called directly from generated code and not on the serve mux.

Reply:
oof, good catch, yes